### ISSUE TITLE
different indexes on pytorch dataloader with trains

### ISSUE THREAD

**Issue by tomeramit** [allegroai/trains-agent#21]
Hi,

I'm launching train with trains, and when I execute different trains I get different indexes from the data loader of pytorch (when I launch it locally, it produce the same indexes) do you have a clue for why this is happening?

attached logs of packages and stuff:

2020-05-28T09:10:58.454Z trains-agent-lv-beast-keras-dl:gpu0,1 INFO task cdb4b8edc4344f078a66c915e0bd6e0f pulled from ca9acd7a32df49e5b4ea41760e252be7 by worker trains-agent-lv-beast-keras-dl:gpu0,1

2020-05-28T09:11:03.497Z trains-agent-lv-beast-keras-dl:gpu0,1 DEBUG Current configuration (trains_agent v0.14.0rc0, location: /tmp/.trains_agent.bpcikde2.cfg):
----------------------
sdk.storage.cache.default_base_dir = ~/.trains/cache
sdk.storage.cache.size.min_free_bytes = 10GB
sdk.storage.direct_access.0.url = file://*
sdk.metrics.file_history_size = 100
sdk.metrics.matplotlib_untitled_history_size = 100
sdk.metrics.images.format = JPEG
sdk.metrics.images.quality = 87
sdk.metrics.images.subsampling = 0
sdk.network.metrics.file_upload_threads = 4
sdk.network.metrics.file_upload_starvation_warning_sec = 120
sdk.network.iteration.max_retries_on_server_error = 5
sdk.network.iteration.retry_backoff_factor_sec = 10
sdk.aws.s3.key = 
sdk.aws.s3.region = 
sdk.aws.boto3.pool_connections = 512
sdk.aws.boto3.max_multipart_concurrency = 16
sdk.log.null_log_propagate = false
sdk.log.task_log_buffer_capacity = 66
sdk.log.disable_urllib3_info = true
sdk.development.task_reuse_time_window_in_hours = 72.0
sdk.development.vcs_repo_detect_async = true
sdk.development.store_uncommitted_code_diff_on_train = true
sdk.development.support_stopping = true
sdk.development.worker.report_period_sec = 2
sdk.development.worker.ping_period_sec = 30
sdk.development.worker.log_stdout = true
agent.worker_id = trains-agent-lv-beast-keras-dl:gpu0,1
agent.worker_name = trains-agent-lv-beast-keras-dl
agent.python_binary = /home/lv-beast/miniconda3/envs/keras-dl/bin/python
agent.package_manager.type = pip
agent.package_manager.pip_version = <20
agent.package_manager.system_site_packages = true
agent.package_manager.force_upgrade = false
agent.package_manager.conda_channels.0 = defaults
agent.package_manager.conda_channels.1 = conda-forge
agent.package_manager.conda_channels.2 = pytorch
agent.venvs_dir = /home/lv-beast/.trains/venvs-builds.2
agent.vcs_cache.enabled = true
agent.vcs_cache.path = /home/lv-beast/.trains/vcs-cache.2
agent.venv_update.enabled = false
agent.pip_download_cache.enabled = true
agent.pip_download_cache.path = /home/lv-beast/.trains/pip-download-cache
agent.translate_ssh = true
agent.reload_config = false
agent.docker_pip_cache = /home/lv-beast/.trains/pip-cache
agent.docker_apt_cache = /home/lv-beast/.trains/apt-cache.2
agent.docker_force_pull = false
agent.default_docker.image = nvidia/cuda
agent.git_user = tomer.amit
agent.default_python = 3.7
agent.cuda_version = 102
agent.cudnn_version = 0
api.version = 1.5
api.verify_certificate = true
api.default_version = 1.5
api.http.max_req_size = 15728640
api.http.retries.total = 240
api.http.retries.connect = 240
api.http.retries.read = 240
api.http.retries.redirect = 240
api.http.retries.status = 240
api.http.retries.backoff_factor = 1.0
api.http.retries.backoff_max = 120.0
api.http.wait_on_maintenance_forever = true
api.http.pool_maxsize = 512
api.http.pool_connections = 512
api.api_server = http://localhost:8008
api.web_server = http://localhost:8080
api.files_server = http://localhost:8081
api.credentials.access_key = LYH4CB09H4VY49TCRT90
api.host = http://localhost:8008
Executing task id [cdb4b8edc4344f078a66c915e0bd6e0f]:
repository = https://bodyvisionmedical.visualstudio.com/LungVision/_git/MachineLearning
branch = develop
version_num = 6ae5d42da77f9fe01c7a09264137bdd147794ba4
tag = 
entry_point = training_script.py
working_dir = LvObjects/Lv3D/TomoGan
Using base prefix '/home/lv-beast/miniconda3/envs/keras-dl'
New python executable in /home/lv-beast/.trains/venvs-builds.2/3.5/bin/python
Installing setuptools, pip, wheel...
done.
Using cached repository in "/home/lv-beast/.trains/vcs-cache.2/MachineLearning.2f27b16d384a01d04c7eedd7180ff087/MachineLearning"
Fetching submodule LungVision

2020-05-28T09:11:08.533Z trains-agent-lv-beast-keras-dl:gpu0,1 DEBUG Note: switching to '6ae5d42da77f9fe01c7a09264137bdd147794ba4'.
You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.
If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:
  git switch -c <new-branch-name>
Or undo this operation with:
  git switch -
Turn off this advice by setting config variable advice.detachedHead to false
HEAD is now at 6ae5d42 Merged PR 3169: add print to dataset index

2020-05-28T09:11:23.578Z trains-agent-lv-beast-keras-dl:gpu0,1 DEBUG url: https://bodyvisionmedical.visualstudio.com/LungVision/_git/MachineLearning
branch: HEAD
commit: 6ae5d42da77f9fe01c7a09264137bdd147794ba4
root: /home/lv-beast/.trains/venvs-builds.2/3.5/task_repository/MachineLearning
Applying uncommitted changes
Collecting pip<20
  Using cached pip-19.3.1-py2.py3-none-any.whl (1.4 MB)
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 20.1.1
    Uninstalling pip-20.1.1:
      Successfully uninstalled pip-20.1.1

2020-05-28T09:11:28.601Z trains-agent-lv-beast-keras-dl:gpu0,1 DEBUG Collecting Cython
  Using cached https://files.pythonhosted.org/packages/57/ac/b095febb27a241a3e09374cc85fb58cecb73439cab59949e0c55b67aac8d/Cython-0.29.19-cp35-cp35m-manylinux1_x86_64.whl
Installing collected packages: Cython
Successfully installed Cython-0.29.19
Running task id [cdb4b8edc4344f078a66c915e0bd6e0f]:
[LvObjects/Lv3D/TomoGan]$ /home/lv-beast/.trains/venvs-builds.2/3.5/bin/python -u training_script.py
Summary - installed python packages:
pip:
- absl-py==0.9.0
- albumentations==0.4.3
- astor==0.8.1
- astra-toolbox==1.9.0.dev0
- attrs==19.3.0
- bvframework==0.1.3
- cachetools==3.1.1
- certifi==2018.8.24
- chardet==3.0.4
- cloudpickle==1.2.2
- colorama==0.4.3
- colorlog==4.1.0
- cycler==0.10.0
- Cython==0.29.19
- dask==2.6.0
- decorator==4.4.1
- dill==0.2.8.2
- elasticdeform==0.4.6
- enum34==1.1.6
- funcsigs==1.0.2
- furl==2.1.0
- future==0.18.2
- gast==0.2.2
- google-api-python-client==1.7.11
- google-auth==1.10.2
- google-auth-httplib2==0.0.3
- google-auth-oauthlib==0.4.1
- google-pasta==0.1.8
- grpcio==1.26.0
- h5py==2.10.0
- httplib2==0.11.3
- humanfriendly==4.18
- idna==2.8
- imageio==2.3.0
- imgaug==0.2.9
- importlib-metadata==1.5.0
- joblib==0.14.1
- jsonmodels==2.4
- jsonpickle==1.2
- jsonschema==3.2.0
- Keras-Applications==1.0.8
- Keras-Preprocessing==1.1.0
- kiwisolver==1.0.1
- llvmlite==0.25.0
- log-utils==0.3.4
- Markdown==3.1.1
- matplotlib==2.2.2
- multiprocess==0.70.5
- networkx==2.4
- nose==1.3.7
- numba==0.40.0
- numpy==1.18.1
- oauth2client==4.1.3
- oauthlib==3.1.0
- olefile==0.46
- opt-einsum==3.1.0
- orderedmultidict==1.0.1
- pandas==0.23.4
- pathlib2==2.3.5
- pathos==0.2.1
- pigar==0.9.2
- Pillow==5.2.0
- pkginfo==1.5.0.1
- plotly==4.5.0
- pox==0.2.7
- ppft==1.6.4.7.1
- protobuf==3.11.2
- psutil==5.6.7
- pyasn1==0.4.8
- pyasn1-modules==0.2.7
- pyclipper==1.0.6
- PyJWT==1.7.1
- pyparsing==2.4.6
- pyrsistent==0.15.7
- pyserialization==0.1.1
- python-dateutil==2.8.1
- pytz==2019.3
- PyWavelets==1.0.1
- PyYAML==3.12
- requests==2.22.0
- requests-file==1.4.3
- requests-oauthlib==1.3.0
- retrying==1.3.3
- rsa==4.0
- scikit-image==0.15.0
- scikit-learn==0.20.0
- scipy==1.1.0
- semantic-version==2.8.2
- Shapely==1.5.16
- SimpleITK==0.10.0
- simplejson==3.16.1
- six==1.14.0
- tabulate==0.8.6
- tensorboard==2.1.0
- tensorflow==2.1.0
- tensorflow-estimator==2.1.0
- termcolor==1.1.0
- toolz==0.10.0
- torch==1.4.0
- torchvision==0.5.0
- tornado==5.1.1
- tqdm==4.42.0
- trains==0.13.2
- typing==3.7.4.1
- uritemplate==3.0.1
- urllib3==1.25.8
- virtualenv==16.7.9
- vtk==8.1.1
- Werkzeug==0.16.1
- wrapt==1.11.2
- zipp==1.1.0

**Comment by bmartinn** [comment]
Hi @tomeramit , I think I'll need a bit more information here.
What could influence your `DataLoader` ?
Basically *trains* will not interfere with the pytorch DataLoader, so this is not directly connected.
That said a few ideas to let us start debugging:
- Is your DataLoader influenced by OS environments ?
- Could it be that you have some configuration file that is not part of the git repo, and you have a local copy on your machine, but since the *trains-agent* clones the repo, it is missing there ? (notice that ignore files, or files that are not added to git will not be part of the "uncommitted changes" of the experiment)

What do you think?

**Comment by tomeramit** [comment]
Hi @bmartinn, thanks for the quick response,
when I launch the training script from windows or Linux, I exactly the same indexes for different runs.

but when I train with trains (execute specific train, and then reset and execute it again from the web interface, exactly same version, and same code) I get different indexes

I set all the seeds  
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False
    np.random.seed(1)
    torch.manual_seed(0)
    random.seed(4)

before I call to Task.init(), maybe this can cause a problem? (the trains launch maybe start from the exact line of the Task init?)

**Comment by tomeramit** [comment]
Is your DataLoader influenced by OS environments ?
Could it be that you have some configuration file that is not part of the git repo, and you have a local copy on your machine, but since the trains-agent clones the repo, it is missing there ? (notice that ignore files, or files that are not added to git will not be part of the "uncommitted changes" of the experiment)

**Comment by bmartinn** [comment]
Hi @tomeramit 
Form what I understand when you say "different indexes" you mean something changes in the Random seed, and that has implications on the way you pull data from the DataLoader, is that correct?
 
If this is the case, Trains does its best to set all random seeds when you call `Task.init`, the specific code can be found [here](https://github.com/allegroai/trains/blob/master/trains/utilities/seed.py#L10)
This means that if you want to have your own seeds I recommend setting them after the `init()` call.


> Update: it happened when I do the same executions from different git commits with trains (for example I execute train, then I add print and commit it, and then I execute again - I get different indexes)

That said, there is no difference in execution between running manually and running with the *trains-agent*, it seems the root cause is that you are actually getting a random seed, and doesn't seem to be related to Trains.  Are you using torch.distributed ?

**Comment by tomeramit** [comment]
Thanks for the answer,
In the end, I couldn't find the root cause of the problem, but it seems that moving the seed part after the init() call made it deterministic.

**Comment by bmartinn** [comment]
@tomeramit I'm happy to hear the issue was solved. 
Anyhow, I think we should add it to the FAQ / documentation, just in case :)

**Comment by tomeramit** [comment]
Agreed, if you recommend setting the seeds after the init it should be mentioned somewhere
